### ISSUE TITLE
Create a migration scheme for Godot Engine resources

### ISSUE THREAD

**Issue by fire** [godotengine/godot-proposals#5945]
### Describe the project you are working on

We are making a 3d social vr online game.

### Describe the problem or limitation you are having in your project

We want to keep distributed scenes updated and migrate to the latest as soon as possible. This solves making it easier to upgrade godot engine scenes and scanning for security problems.

@SaracenOne was worried about code improvements when we do a possible compatibility breakage like in https://github.com/godotengine/godot/pull/65312 where an enumeration had an entry removed and then a new entry was added in the same spot. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

We want to use the original resource. Extract the mapping of property name to Variant. Run the migration by taking Map<StringName, Variant>, the two pairs of versions, and migrations. Save the properties into a MissingResource. On load the MissingResource will convert to a resource.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

1. Resource -> MissingResource-> Resource [https://docs.godotengine.org/en/latest/classes/class_missingresource.html](https://docs.godotengine.org/en/latest/classes/class_missingresource.html?highlight=missingresource)
2. For each Resource property: get_property_list(), get(), set() are migrated. original_class and is_recording_properties properties exist and are set in the MissingResource.
3. MissingResource can be checked easily for security and complexity explosions via a conversion to HashMap<StringName, Variant> and because MissingResource has less implementation to verify.
4. In HashMap format check for problems and run schema migrations. (Note that you can convert the HashMap to json in a straightforward manner.)
5. When you load a MissingResource it will auto-convert to the original class.

# Reference:

* https://github.com/godotengine/godot/pull/60597
* https://github.com/godotengine/godot/blob/f5978f4d97150d9e7a1598f9637f0f3c06d72d33/core/io/missing_resource.h
* https://github.com/bruvzg/gdsdecomp/blob/master/compat/resource_loader_compat.cpp

# Contributors:

* @fire 
* @nikitalita
* @lyuma 

### If this enhancement will not be used often, can it be worked around with a few lines of script?

We need to store the migration in the engine.

### Is there a reason why this should be core and not an add-on in the asset library?

The engine developers need to create a migration for any compatibility breaking changes.

**Comment by Creta5164** [comment]
I agree that this feature is needed, my current situation is different, but I'm migrating in the following way for now.

![Image](https://github.com/user-attachments/assets/0bd4174b-f4b5-45b1-bcb7-cc681edb9947)

Even when I was using Unity, most of the time to do such migrations, I had to create two different types of “old” and “new” properties, migrate them to the new ones, and then replace the “new” ones with the “old” ones again after a period of time.
Personally, I think this is pretty annoying, and if the feature is built as the author says it will be, I think it will be quite useful to solve this annoyance.

**Comment by Creta5164** [comment]
I'm not sure I'm qualified to comment here, but if the workaround is to simply handle the old/current behavior with just one bool property, I wonder if we'd have the following issue :

- Change property to different type
- Additional properties for migration

So I'm thinking, as you suggested in your suggestion, to import the existing resource into an intangible instance called MissingResource and resolve its contents.
(I was thinking get them into varint dictionary)
Please correct me if my understanding of the concept is wrong.

I'm also curious how other people (especially add-on and core engine developers) deal with migrations in these situations, I've only been writing Godot for a short time, so I'm afraid that I may not even realize that there is a much better way to deal with these migration issues.

**Comment by fire** [comment]
Both MissingResource and MissingNode already exist.

https://docs.godotengine.org/en/stable/classes/class_missingresource.

Although the team has yet to agree to this proposal of creating a change-by-change migration, it is the best solution, as far as I know.

This proposal is in limbo.

**Comment by Creta5164** [comment]
I just found more active proposal issue(https://github.com/godotengine/godot-proposals/issues/10644) that I think possibly related about this problem.

**Comment by Creta5164** [comment]
I want to share my way about doing migration just in case if anyone reached here by searching.

In #10644, @AThousandShips said, `_set` can grab serialized property and values so I managed to do migrate process.

Here's node or resource's part of code.

```gdscript
## Handles version control that makes old version up to date.
var _migration_context := MigrationContext.new(
        self,
        _on_migration_finished
    )

func _on_migration_finished() -> void:
    
    #TODO: Write finalization logic here for after migration if necessary.
    
    _migration_context = null
    
    pass

func _set(property: StringName, value: Variant) -> bool:
    
    if _migration_context != null:
        _migration_context.queue_property(property, value)
    
    return false
```

Here's inline class code for handling migration process.

```gdscript
class MigrationContext extends RefCounted:
    
    var context: DreamRiggerSprite    #It's my node type.
    
    var queued_values:    Dictionary[StringName, Variant]
    var is_queue_started: bool
    
    var _finish_callback: Callable
    
    func _init(context: DreamRiggerSprite, finish_callback: Callable) -> void:
        
        self.context     = context
        _finish_callback = _finish_callback
        
        pass
    
    func queue_property(property_name: StringName, value: Variant) -> void:
        
        set_property(property_name, value)
        
        if !is_queue_started:
            
            is_queue_started = true
            _process.call_deferred()
        
        pass
    
    func _process() -> void:
        
        if !is_instance_valid(context):
            return
        
        if context.version >= DreamRiggerMetadata.VERSION:
            
            context._migration_context = null
            return
        
        while context.version < DreamRiggerMetadata.VERSION:
            
            _migrate(context.version)
            context.version += 1
            
            pass
        
        for property_name: StringName in queued_values:
            
            if property_name == &"version":
                continue
            
            context.set(property_name, queued_values[property_name])
            
            pass
        
        if is_instance_valid(_finish_callback):
            _finish_callback.call()
        
        context._migration_context = null
        
        pass

    func _migrate(current_version: int) -> void:
        
        match DreamRiggerMetadata.to_semantic_version(current_version):   #It simply converts number to readable semantic version.
            
            #TODO: Write migration logic here if necessary.
            
            &"alpha-1.0.0":
                
                move_property(&"pivot", &"offset")
                
                pass
            
            pass
        
        pass
    
    func get_property(name: StringName) -> Variant:
        
        return queued_values[name]\
            if queued_values.has(name)\
          else null
    
    func set_property(name: StringName, value: Variant) -> void:
        
        queued_values[name] = value
        
        pass
    
    func move_property(from: StringName, to: StringName) -> void:
        
        set_property(to, delete_property(from))
        
        pass
    
    func delete_property(name: StringName) -> Variant:
        
        var value := get_property(name)
        queued_values.erase(name)
        
        return value
    
    pass
```

Here's `DreamRiggerMetadata` if you want to know about it in `_migrate` method.

```gdscript
class_name DreamRiggerMetadata

const VERSION: int = 2

const SEMANTIC_VERSION: StringName = &"beta-1.0.0"

#TODO: Should be auto-generated.
static func to_semantic_version(version: int) -> StringName:
    
    match version:
        1: return &"alpha-1.0.0"
        
        #Marking it latest, requires add previous VERSION with-
        #semantic version name before this if increases.
        VERSION: return SEMANTIC_VERSION
        
        pass
    
    return &"UNKNOWN"
```
